### PR TITLE
UserInterfaceHook in ILIAS 6.0

### PR DESCRIPTION
**Pull request by Amstutz**
Dear ILIAS Developer

Thank you a lot for checking out this PR and joining the discussion to make it better.

TDL: 
This PR adds one method to simplify extending UI Components in ILIAS 6.0 with plugins. Further this PR contains various comments to improve the documentation around the UserInterfaceHook Pluginslot. This PR is shipped with an [example plugin](https://github.com/Amstutz/UIHookDemo/), showing the various ways of how the UI ILIAS 6.0 might be extended by the usage of plugins. 

Goal:
- This PR aims to deepen and give some flesh to the discussion on how to extend the UI of ILIAS 6.0 on forwards.

Background: 
In the last Page Layout Revision Joure Fixe we discussed that several options of how to use the User Interface Hook Plugin that used to work up to ILIAS 5.4 might be unusable due to the migration towards Global Screen and UI Components. We further observed that the current usage of the slot is highly diverse and in several cases probably not the best way to tackle certain problems but rather an „everything is doable“ option to cover all demands that can not be served in any other way. There seemed to be a consensus that we can not simply cut this option, but rather elaborate how several of the use cases can be tackled by using the new options given in ILIAS 6.0. @klees and @amstutz in their role as UI Components Coordinators agreed to shoulder the task to add some „sugar“ to simplify the extensibility of UI components by plugin slots. Since @rklees currently shoulders a great number of tasks in the UI Component @amstutz worked out a proposal for this. Note that this is not yet reviewed by @klees. Further note that this PR is made in the role as UI Components Coordinator and not in the role of a member of the Technical Board.

Contents:
- Added afterClientInitialization for Plugins in Initialization: This hook gives every plugin the possibility to (re)-wire parts of the initialization. Due to the service locator nature of our way of doing DI in ILIAS, this comes with the serious drawback, that different plugins might mess with the re-wiring of others if one does not take care of keeping the context and the re-wiring as slim as possible. Note that such conflicts did obviously exist before, when different plugins where hacking into the same html output. Without this method it is hard to find a good place to perform the re-wiring, since either the DIC is not fully loaded, or already did perform some work. Also note that such re-wiring does provide IMO much superior ways of extending the UI (e.g. one does not need to render output which is not really needed etc.). Have a look a the example at: https://github.com/Amstutz/UIHookDemo/ to find different methods of extending the UI in 6.0
- The most lines changed try to improve the documentation of the slot by moving the documentation closer to the code and by adding a readme for the UI components plugin slot. 
- Note that this PR proses to deprecate the methods getHTML and modifyGUI of ilUIHookPluginGUI. The reasoning behind this is given in the code.
- Few lines to improve doc on Global Screen

Example:
Important to quickly show, how this slot can be used from ILIAS 6.0 on forward, I propose the following example to be added to the list of example plugins of ILIAS and to be linked in the readme: https://github.com/Amstutz/UIHookDemo. It gives an overview of the various ways the slot can be used.

Discussion:
I thought for quite some time, on how to add some additional sugar to simplify the extensibility of the UI Components. All I did was adding the afterClientInitialization for plugins. IMO the UI Components are by Design extensible and highly flexible. So IMO there is not much more needed there. However, I am still not sure if I registered all the use cases. There might be pieces missing, so please join discussion on how to make ILIAS 6.0 as easily extensible as possible.

Thx a lot 
@amstutz

### CODE REVIEW

**Review by alex40724 — CHANGES_REQUESTED** [review]
Hi Timon,

thank you very much for working out a proposal.

I will give up maintainership for the UI Hook plugin slot, so I do not give a review under this role any more. I understand that the HTML based hook should be deprecated.

I am also listed as implicit maintainer for the Services/Init directory, and I would like to give opinion on this part, even if it is at least equally related to the maintainership of src/DI (which currently only lists @chfsx  as implicit maintainer).

I feel that opening the DI for replacement of components this way in plugins may not be the best way to move forward here.

- This interface is not UI related. It is acting on the DI directly and everything served by the DI can be replaced.
- The example replaces the complete renderer on the highest level possible. Promoting this as a good practive may result in lots of Plugins that really try to solve things this way. The old "interface" allowed to modify very distinct low level UI parts, e.g. based on the template being served. Replacing the whole renderer for these purposes seems even more error-prone to me. I fear if more plugins act on high level components the probability of conflicts may increase.

I would prefer a solution that

- Only adresses the UI level of the application.
- More explicitly targets distinct components on distinct screens. The help systems uses screen IDs internally, we may try to make this a concept of the UI framework. This is highly related to the Global Screen Context discussion imo.

**Comment by Amstutz** [comment]
Hi @alex40724 

Thanks a lot for the feedback.

> I feel that opening the DI for replacement of components this way in plugins may not be the best way to move forward here.

Full ack. It is the option for use cases in which a huge extend of flexibility is needed. Note that this opening is not only usable for UI Plugins but all. This comes with a very serious drawback, as pointed out at several places: Plugins might run into conflicts.

> The example replaces the complete renderer on the highest level possible. Promoting this as a good practive may result in lots of Plugins that really try to solve things this way. The old "interface" allowed to modify very distinct low level UI parts, e.g. based on the template being served. Replacing the whole renderer for these purposes seems even more error-prone to me. I fear if more plugins act on high level components the probability of conflicts may increase.

Only one example does this, the JsonRenderer, which is described as the "Ruthless Solution". It is possible. It might be useable in some edge cases. I do not think that many devs. will go through the pain of writting an independent renderer. Did you check out the custom Metabar Example. This does more or less what you are asking for. Just extending the renderer for the Metabar and there add a few minor details: https://github.com/Amstutz/UIHookDemo/tree/master/src/DIC/CustomMetaBar.

Does this also go to far?

Further I hope that more components will be able to be extended as the main menu via global screen. Did you check out the main menu example https://github.com/Amstutz/UIHookDemo/tree/master/src/MainMenu? What do you think of that?

I agree that exchange parts of the DIC should be reserved for the hard nuts of available contexts. However I do not believe it to be evil to provide plugins with a hook enabling exactly that: Have a controlled way of interfering with ILIAS just after the initialization is complete. Up to now I do not think that such an option does exist. 

> More explicitly targets distinct components on distinct screens. The help systems uses screen IDs internally, we may try to make this a concept of the UI framework. This is highly related to the Global Screen Context discussion imo.

I agree and we should continue this discussion. I also believe that the global screen is the way forward. However, this can be the insurance that almost all szenarios still have a patch free solution for ILIAS 6.0.

Thanks a lot for opening the discussion. As always you feedback is very welcome and I will put some more thought into this. Hopefully I'll we get more feedback for this IMO important topic.

**Review by klees — COMMENTED** [review]
Hi everybody!

First of all: thx @Amstutz for adding some documentation. And also for working out this proposal.

This proposal and the examples showcase the power of dependency injection. After all, DI helps to build a truly modular system, which in turn also means that parts of the system can be extended, exchanged or maybe even disabled. This, after all, is what all plugins are designed to do: They are supposed to extend, exchange or disable functionality of standard ILIAS. So after all, using DI in conjunction with plugins to me feels natural and obvious in general. I'm not sure, though, if we are there yet with the DI in ILIAS.

As @Amstutz correctly points out, the DI container currently is mostly used
as service locator. This means, that in fact some code might already have requested some dependency from the DI at the point where a plugin interferes. This means, that we might end up with a situation where one part of the code uses X to resolve some dependency and the other part of the code uses Y for the same dependency after a plugin exchanged X for Y. This will lead to super strange bugs. We could say: Well, that's on plugin developers. But at the end of the day, this will spawn tickets in our tracker which someone has to triage.

Another problem with the current state of DI in ILIAS imo is, that we only have little definitions for the interfaces that some dependency is supposed to implement. This means that a dependency could be exchanged in the DI. But how does the replacement need to look like? Again, we could shift that responsibility to the plugin developers, but we will get tickets.

As @Amstutz also correctly points out, though, this situation does not differ that much from the current situation with the UI-Hook-Plugin in that regard (where plugins interfere at html-level), but the surface for related problems will become much bigger (whole system instead of html only).

For the time being I would prefer to settle on a middle ground. Allowing the UI-Hook-Plugin to replace the (or some) services provided by the UI-Framework will not suffer from the two general problems pointed out above that much. We have defined interfaces and we indeed do proper DI (and not service location) for these components. I would thus propose to introduce a method like this to the UI-Hook-Plugin:

```php
public function getRenderer(\ILIAS\UI\Renderer $original) :  \ILIAS\UI\Renderer {
	...
}
```

People could use that to decorate the original renderer with some special logic. Maybe we also want to replace more specific services from the UI-framework like `ui.component_renderer_loader`, but imo we would need to publish the according interface then (currently its internal to the UI-framework). This would make replacing specific renderers for specific components even simpler.

This would also solve two issues from @alex40724, namely:

> This interface is not UI related. It is acting on the DI directly and everything served by the DI can be replaced.

and

> I would prefer a solution that [o]nly adresses the UI level of the application.

This might leave this issue open:

> The example replaces the complete renderer on the highest level possible. Promoting this as a good practive may result in lots of Plugins that really try to solve things this way. The old "interface" allowed to modify very distinct low level UI parts, e.g. based on the template being served. Replacing the whole renderer for these purposes seems even more error-prone to me. I fear if more plugins act on high level components the probability of conflicts may increase.

I'm not sure, if this also would be an issue with the proposed solution. After all, I could just decorate the original renderer and pick the component that I'm interested in:

```php
public function getRenderer(\ILIAS\UI\Renderer $original) :  \ILIAS\UI\Renderer {
	return new class ($original) implement \ILIAS\UI\Renderer {
		protected $original;
		public function __construct($original) {
			$this->original = $original;	
		}
		public function render($component) {
			if ($component instanceof \ILIAS\UI\Component\Button\Button) {
				$html = $this->original->render($button);
				// do whatever
				return $html;
			}
			return $this->original->render($button);
		}
		public function renderAsync($component) {
			return $this->original->renderAsync($component);
		}
		public function withAdditionalContext(Component $context) {
			$this->original->withAdditionalContext($context);
			return $this;
		}
	};
}
```

Since we move away from templates for individual modules or services, it will indeed be harder to pick "the right" component here. This is not a problem of the proposed mechanism, though, but a result of the decision to move UIs of all modules and services to some common set of UI-components and get rid of individual templates for services or modules long term.

As @alex40724 points out, we could ease to pick "the right" component via something like screen-IDs. I'm not sure if I see this as a problem of the UI-framework (as we understand it now). Currently, the UI-framework covers the "how should it look" part of the UI. This asks for "where is the user in the system". Currently, via the templates in the services and modules, these concerns overlap. But we seem to be working towards a system, where the concerns are orthagonal. To me this again points at the core problem of the UI-Hook-Plugin: It's used for too much fairly disparate stuff. I'm sure we will be finding more satisfying solutions for its use cases. And also: We should have a discussion if and how we want ILIAS to be extendable in the future soon.

Best regards!

**Comment by Amstutz** [comment]
Hi @klees 

The "middle-ground" solution would be fine with me. I am not a big advocate of the handing our the complete DIC at the current stage to the plugin.

However, I it would be neat to also have a defined way to replace components as well. So something like getUIComponentFactory($original) would be neat as well. IMO I would probably use this more often than just replacing the renderer. Note how this works nicely with a custom renderer as shown in one of the examples.

Greetings.

**Comment by Amstutz** [comment]
We discussed this PR in our weekly meeting concerning ILIAS 6.0. We agreed that there are still further is needed to decide anything on this. Again concerns have been void against exchanging parts of the DI by plugins.

Dear fellow devs: I believe the midde-ground solution proposed by richard  along with the promise of trying to shift more extension logic towards the global screen could be a good way to go for ILIAS 6.0. Note that without this hook into init, we still depend on rendering everything first and have a very late option to interfere with that later. By providing custom renderers for certain components in certain context, one has a much better handle on this.

File: Services/Component/classes/class.ilPlugin.php
Comment (chfsx) [inline Services/Component/classes/class.ilPlugin.php]: As others have already mentioned, this is certainly a very powerful but also dangerous method. Certainly also because a lot is not clearly defined within the DIC, there are no interfaces etc..

File: src/GlobalScreen/Scope/MainMenu/Collector/Renderer/TypeRenderer.php
Comment (chfsx) [inline src/GlobalScreen/Scope/MainMenu/Collector/Renderer/TypeRenderer.php]: Thanks, I forgot to add something here...

**Review by chfsx — CHANGES_REQUESTED** [review]
Many thanks @Amstutz for this suggestion and also the discussion so far! I also see it that the way via `afterClientInitialization` is more dangerous, but not less dangerous than before. global $DIC could always be modified by plugins, but not by such a offered interface. I like @klees' suggestion to break this down to the UI part for now and offer the possibility to replace a renderer.

Furthermore, I already suggested a supplement for the UIHook plugins in PR #1936, which would complement each other well.

**Comment by Amstutz** [comment]
Thanks to you all for the feedback. I agree with that afterClientInitialization is quite error prone. 

> global $DIC could always be modified by plugins, but not by such a offered interface. 

IMO, no, this is the reason we need this for exchange the UI Renderer AND/OR components in the first place. If not done in the initialization but later, there is a risk that the component to be replaced is already in-use and did work. So IF you want to replace something, after INIT is correct, otherwise you can just cross your fingers and hoper for the best.

I strongly believe we need the following two things to really provide the needed flexibility for 6.0:

- Replace factories of UI-Components. This is very handy to extend components/factories. I provided an example in the linked repo providing a show case, which makes to power of such a thing transparent. 
- Replace the renderer of UI-Components

@klees , @alex40724 and @chfsx would you agree with those two measures for 6.0.

@chfsx : Just browsed through #1936, looks awsome. I believe with your proposal on how global screen components can be exchanged in a very controled manner, the above two options to replace any other component if needed and the [navigational context ](https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/src/NavigationContext)(hoping a readme for this is also in the pipeline ;-) ). We are in a much better shape when it comes to customizing UI then before 6.0.

**Comment by chfsx** [comment]
@Amstutz 

> @klees , @alex40724 and @chfsx would you agree with those two measures for 6.0.

I'd be absolutely fine!

> We are in a much better shape when it comes to customizing UI then before 6.0.
I agree, the mix of this PR and the changes in 1936 will be a huge step for the UI-Hook-Plugins

**Comment by klees** [comment]
> would you agree with those two measures for 6.0.

Yes, this is fine for me.

**Comment by alex40724** [comment]
> @klees , @alex40724 and @chfsx would you agree with those two measures for 6.0.

Yes, thanks.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 01 JUL 2019 : We discussed Timon's suggestions and accept to set the two methods getHTML and modifyGUI of ilUIHookPluginGUI as deprecated and to introduce a hook after initialisation of UI components to exchange factories and renderers of UI components. Please modify the PR accordingly and set it on the agenda again for a final decision.

**Comment by Amstutz** [comment]
Adapted according to the JF decision. afterClientInitialization was replaced by exchangeUIRendererAfterInitialization and exchangeUIFactoryAfterInitialization which are called after the initialisation of the UI Framework in ilIInitlialisation.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 12 AUG 2019 : We highly appreciate the mofified PR and accept it for ILIAS 6. As already announced by Alexander Killing, he wants to handle over the maintenance of this plugin slot to Fabian Schmid (who accepted this). Please update the entry for maintenance in maintenance.md.

**Comment by Amstutz** [comment]
@mjansenDatabay : Thx for your support in fixing the tests!

@chfsx : This is ready to be merged, I leave this to you, since this is now your component. Thx a lot.

**Review by chfsx — APPROVED** [review]

**Pull request merged**
